kyaml/fn/framework/command/command.go

```diff
@@ -0,0 +1,156 @@
+// - The argument "-", if present, will cause resources to be read from STDIN as well.
+// The output will be a raw stream of resources (not wrapped in a List type).
+// Example usage: `cat input1.yaml | go run main.go config.yaml input2.yaml input3.yaml -`
+//
+// If `disableStandalone` is set to true, all arguments are ignored, and STDIN must contain
+// a Kubernetes List type. To pass a function config in this mode, use a ResourceList as the input.
+// The output will be of the same type as the input (e.g. ResourceList).
+// Example usage: `cat resource_list.yaml | go run main.go`
+//
+// By default, any error returned by the ResourceListProcessor will be printed to STDERR.
+// Set noPrintError to true to suppress this.
+func Build(p framework.ResourceListProcessor, disableStandalone, noPrintError bool) *cobra.Command {
+	cmd := cobra.Command{}
+
+	addGenerateDockerfile(&cmd)
+
+	var printStack bool
+	cmd.Flags().BoolVar(&printStack, "stack", false, "print the stack trace on failure")
+	cmd.Args = cobra.MinimumNArgs(0)
+	cmd.SilenceErrors = true
+	cmd.SilenceUsage = true
+
+	cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		var readers []io.Reader
+		rw := &kio.ByteReadWriter{
+			Writer:                cmd.OutOrStdout(),
+			KeepReaderAnnotations: true,
+		}
+
+		if len(args) > 0 && !disableStandalone {
+			// Don't keep the reader annotations if we are in standalone mode
+			rw.KeepReaderAnnotations = false
+			// Don't wrap the resources in a resourceList -- we are in
+			// standalone mode and writing to stdout to be applied
+			rw.NoWrap = true
+
+			for i := range args {
+				// the first argument is the resourceList.FunctionConfig
+				if i == 0 {
+					var err error
+					if rw.FunctionConfig, err = functionConfigFromFile(args[0]); err != nil {
+						return errors.Wrap(err)
+					}
+					continue
+				}
+				if args[i] == "-" {
+					readers = append([]io.Reader{cmd.InOrStdin()}, readers...)
+				} else {
+					readers = append(readers, &deferredFileReader{path: args[i]})
+				}
+			}
+		} else {
+			readers = append(readers, cmd.InOrStdin())
+		}
+		rw.Reader = io.MultiReader(readers...)
+
+		err := framework.Execute(p, rw)
+		if err != nil && !noPrintError {
+			_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "%v", err)
+		}
+		// print the stack if requested
+		if s := errors.GetStack(err); printStack && s != "" {
+			_, _ = fmt.Fprintln(cmd.ErrOrStderr(), s)
+		}
+		return err
+	}
+
+	return &cmd
+}
+
+// addGenerateDockerfile adds a "gen" subcommand to create a Dockerfile for building
+// the function as a container.
+func addGenerateDockerfile(cmd *cobra.Command) {
+	gen := &cobra.Command{
+		Use:  "gen [ROOT]",
+		Args: cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return ioutil.WriteFile(filepath.Join(args[0], "Dockerfile"), []byte(`FROM golang:1.15-alpine as builder
+ENV CGO_ENABLED=0
+WORKDIR /go/src/
+COPY . .
+RUN go build -v -o /usr/local/bin/function ./
```

Comment:
I think this might need to use:

`-tags netgo -ldflags '-w'`

`-tags netgo`: forces using the built-in net package which is missing in alpine AFAIK
`-ldflags '-w'`: disables debug so the output is smaller

And then either use `CGO_ENABLED=0` or, if you want to allow people to use cgo, make sure you make `ldflags` be `-ldflags '-w -extldflags "-static"'` so the linked C part is also static and added to the binary.

source: https://medium.com/@diogok/on-golang-static-binaries-cross-compiling-and-plugins-1aed33499671